### PR TITLE
Ocrvs 2877 Form Config BE

### PR DESCRIPTION
**Pull request by sadmananik**

### CODE REVIEW

File: packages/metrics/src/influxdb/client.ts [hunk 1]
```diff
@@ -187,6 +187,29 @@ export const query = <T = any>(q: string): Promise<T> => {
...
+      influx.dropMeasurement('death_reg', INFLUX_DB),
+      influx.dropMeasurement('in_complete_fields', INFLUX_DB),
+      influx.dropMeasurement('declaration_time_logged', INFLUX_DB),
+      influx.dropMeasurement('declaration_event_duration', INFLUX_DB),
+      influx.dropMeasurement('certification_payment', INFLUX_DB),
+      influx.dropMeasurement('correction_payment', INFLUX_DB),
+      influx.dropMeasurement('declarations_started', INFLUX_DB),
+      influx.dropMeasurement('declarations_rejected', INFLUX_DB)
+    ]).then(() => {
+      return {
```
Comment (Zangetsu101) [inline packages/metrics/src/influxdb/client.ts:203]: I see two `return` statements here. Was this intentional?

File: packages/workflow/src/constants.ts [hunk 2]
```diff
@@ -13,6 +13,8 @@ export const HOST = process.env.HOST || '0.0.0.0'
 export const PORT = process.env.PORT || 5050
 export const HEARTH_URL = process.env.HEARTH_URL || 'http://localhost:3447/fhir'
 export const OPENHIM_URL = process.env.OPENHIM_URL || 'http://localhost:5001'
+export const APPLICATION_CONFIG_URL =
+  process.env.APPLICATION_CONFIG_URL || 'http://localhost:2021/'
```
Comment (Zangetsu101) [inline packages/workflow/src/constants.ts:17]: This probably also needs to be added on compose file

File: packages/gateway/src/features/formDraft/root-resolvers.ts [hunk 3]
```diff
@@ -79,22 +137,36 @@ function createOrUpdateFormDraftPayload(
...
 }
 
 enum DraftStatus {
   DRAFT = 'DRAFT',
-  PREVIEW = 'PREVIEW',
+  IN_PREVIEW = 'IN_PREVIEW',
   PUBLISHED = 'PUBLISHED',
-  FINALISED = 'FINALISED'
+  DELETED = 'DELETED'
 }
```
Comment (Zangetsu101) [inline packages/gateway/src/features/formDraft/root-resolvers.ts:160]: These wouldn't have been needed if they had been added on schema. These will also have effect on `client` so some amends are needed there too.

**Review by Zangetsu101 — CHANGES_REQUESTED** [review]

File: packages/config/src/handlers/formDraft/deleteFormDraft/handler.ts [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    if (hasTestExtensionOnAllTasks) {
+      //deleting all elastic and influx data
+      try {
+        const token = request.headers.authorization.replace('Bearer ', '')
+        Promise.all([
+          await fetch(`${SEARCH_URL}elasticIndex`, {
+            method: 'DELETE',
+            headers: {
+              'Content-Type': 'application/json',
+              Authorization: `Bearer ${token}`
```
[the remark this answers is not in the document]
Comment (sadmananik) [reply packages/config/src/handlers/formDraft/deleteFormDraft/handler.ts:93]: I was like this in another place. I think to ensure Bearer is always there.

File: packages/gateway/src/features/formDraft/root-resolvers.ts [hunk 5]
```diff
@@ -79,22 +137,36 @@ function createOrUpdateFormDraftPayload(
...
+function modifyFormDraftStatusPayload(
+  formDraft: GQLFormDraftStatusModify
+): IModifyDraftStatusPayload {
+  const formDraftStatusPayload: IModifyDraftStatusPayload = {
+    event: formDraft.event as Event,
+    status: formDraft.status as DraftStatus
+  }
+
+  return formDraftStatusPayload
+}
```
[the remark this answers is not in the document]
Comment (sadmananik) [reply packages/gateway/src/features/formDraft/root-resolvers.ts:149]: This casting was done in two handlers that's why did it on the function to reuse casting code.

File: packages/config/src/handlers/formDraft/createOrupdateFormDraft/handler.ts [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+  comment: Joi.string(),
+  deleted: Joi.array().items(Joi.string())
+})
+
+export const modifyFormDraftStatus = Joi.object({
+  event: Joi.string()
+    .valid(...validEvent)
+    .required(),
+  status: Joi.string()
+    .valid(DraftStatus.IN_PREVIEW, DraftStatus.PUBLISHED)
```
[the remark this answers is not in the document]
Comment (sadmananik) [reply packages/config/src/handlers/formDraft/createOrupdateFormDraft/handler.ts:264]: Yeah. But DELETED will be done by Delete handler and that validation is [here](https://github.com/opencrvs/opencrvs-core/blob/c8840d4cfd6b2d58cec268bcdc4910cbd92dcc08/packages/config/src/handlers/formDraft/deleteFormDraft/handler.ts#L175)

File: packages/metrics/src/influxdb/client.ts (hunk 1 above)
Comment (sadmananik) [reply packages/metrics/src/influxdb/client.ts:203]: It was a mistake. Rmoved return before `Promise.all([`

File: packages/workflow/src/constants.ts (hunk 2 above)
Comment (sadmananik) [reply packages/workflow/src/constants.ts:17]: Yes. good catch.

File: packages/gateway/src/features/formDraft/root-resolvers.ts (hunk 3 above)
Comment (sadmananik) [reply packages/gateway/src/features/formDraft/root-resolvers.ts:160]: I have already changed these statuses. IDK why these appear on PR!

File: packages/metrics/src/influxdb/client.ts (hunk 1 above)
Comment (sadmananik) [reply packages/metrics/src/influxdb/client.ts:203]: But I think i can use await.

File: packages/config/src/handlers/formDraft/createOrupdateFormDraft/handler.ts [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+    //create/update/delete questions
+    if (!isEmpty(questionsDraft.questions)) {
+      const existingQuestions: IQuestionModel[] = await Question.find({
+        fieldId: eventRegex
+      }).exec()
+
+      const allQuestion = partition(questionsDraft.questions, (question) => {
+        return find(existingQuestions, { fieldId: question.fieldId })
+      })
```
[the remark this answers is not in the document]
Comment (sadmananik) [reply packages/config/src/handlers/formDraft/createOrupdateFormDraft/handler.ts]: Yes. Thanks.

File: packages/gateway/src/features/formDraft/root-resolvers.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Zangetsu101) [reply packages/gateway/src/features/formDraft/root-resolvers.ts:149]: Let me rephrase my question. Why do we need the type casting?

File: packages/gateway/src/features/formDraft/schema.graphql [hunk 8]
```diff
@@ -58,16 +58,22 @@ input QuestionInput {
...
+}
+
 type Query {
   getFormDraft: [FormDraft!]
 }
 
 type Mutation {
   createOrUpdateFormDraft(formDraft: FormDraftInput!): FormDraft
+  modifyDraftStatus(formDraft: FormDraftStatusModify!): FormDraft
+  deleteFormDraft(formDraft: FormDraftStatusModify!): FormDraft
```
Comment (Zangetsu101) [inline packages/gateway/src/features/formDraft/schema.graphql:78]: We really don't need two separate Mutations for this. As we can already see the inputs are also same

File: packages/config/src/handlers/formDraft/deleteFormDraft/handler.test.ts [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+//     maxLength: 32,
+//     fieldName: 'myField',
+//     fieldType: 'TEXT',
+//     preceedingFieldId: 'myPreviousFieldId',
+//     required: true,
+//     enabled: true,
+//     custom: true,
+//     initialValue: 'myValue'
+//   }
+// ]
```
Comment (Zangetsu101) [inline packages/config/src/handlers/formDraft/deleteFormDraft/handler.test.ts:77]: There seems to be some comments left

Comment (sadmananik) [reply packages/config/src/handlers/formDraft/deleteFormDraft/handler.test.ts]: Deleted the file
